src/pip/_internal/index/sources.py

```diff
@@ -36,6 +44,57 @@ def _is_html_file(file_url: str) -> bool:
     return mimetypes.guess_type(file_url, strict=False)[0] == "text/html"
 
 
+class _FlatDirectoryToUrls:
+    """Scans directory and caches results"""
+
+    def __init__(self, path: str) -> None:
+        self._path = path
+        self._page_candidates: Optional[List[str]] = None
+        self._file_links_and_project_names: Optional[List[Tuple[str, str]]] = None
+
+    def _scan_directory(self) -> None:
+        """Scans directory once and populates both page_candidates
+        and file_links at the same time
+        """
+        self._page_candidates = []
+        self._file_links_and_project_names = []
+        for entry in os.scandir(self._path):
+            url = path_to_url(entry.path)
+            if _is_html_file(url):
+                self._page_candidates.append(url)
+                continue
+
+            # File must have a valid wheel or sdist name,
+            # otherwise not worth considering as a package
+            try:
+                project_filename = parse_wheel_filename(entry.name)[0]
+            except (InvalidWheelFilename, InvalidVersion):
+                try:
+                    project_filename = parse_sdist_filename(entry.name)[0]
+                except (InvalidSdistFilename, InvalidVersion):
+                    continue
+
+            self._file_links_and_project_names.append((url, project_filename))
+
+    @property
+    def page_candidates(self) -> List[str]:
+        if self._page_candidates is not None:
+            return self._page_candidates
+
+        self._page_candidates = []
```

Comment:
I'm a little confused by this pattern of `self._page_candidates = []` then call `_scan_directory()` which starts with `self._page_candidates = []`. I feel like it would be less surprising to be in one location _or_ the other (but I'm not sure which makes more sense).

Reply:
What about a boolean for `_is_initialized` or similar, and initializing `_page_candidates` as an empty list in the `__init__`? I think pylint/mypy/others should be fine with that and probably conveys intent better than checking if the attribute is None.

Reply:
Implemented